build.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright (C) 2021-2022 The Nitrocli Developers
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+use std::process::Command;
+
+use anyhow::bail;
+use anyhow::Context;
+use anyhow::Result;
+
+const GIT: &str = "git";
+
+/// Format a git command with the given list of arguments as a string.
+fn git_command(args: &[&str]) -> String {
+  args.iter().fold(GIT.to_string() + " ", |mut cmd, arg| {
+    cmd += &(" ".to_owned() + arg);
+    cmd
+  })
+}
+
+/// Run git with the provided arguments and read the output it emits.
+fn git_output(args: &[&str]) -> Result<String> {
+  let git = Command::new(GIT)
+    .args(args)
+    .output()
+    .with_context(|| format!("failed to run `{}`", git_command(args)))?;
+
+  if !git.status.success() {
+    let code = if let Some(code) = git.status.code() {
+      format!(" ({})", code)
+    } else {
+      String::new()
+    };
+
+    bail!(
+      "`{}` reported non-zero exit-status{}",
+      git_command(args),
+      code
+    );
+  }
+
+  let output = String::from_utf8(git.stdout).with_context(|| {
+    format!(
+      "failed to read `{}` output as UTF-8 string",
+      git_command(args)
+    )
+  })?;
+
+  Ok(output)
+}
+
+/// Run git with the provided arguments and report the status of the
+/// command.
+fn git_run(args: &[&str]) -> Result<bool> {
+  Command::new(GIT)
+    .args(args)
+    .status()
+    .with_context(|| format!("failed to run `{}`", git_command(args)))
+    .map(|status| status.success())
+}
+
+/// Create a suffix to add to the regular cargo reported program
+/// version, which includes information about the git commit, if
+/// available.
+fn version_suffix() -> Result<String> {
+  // As a first step we check whether we are in a git repository and
+  // whether git is working to begin with. If not, we can't do much; yet
+  // we still want to allow the build to continue, so we merely print a
+  // warning and continue without a version suffix. But once these
+  // checks are through, we treat subsequent failures as unexpected and
+  // fatal.
+  match git_run(&["rev-parse", "--git-dir"]) {
+    Ok(true) => (),
+    Ok(false) => {
+      println!("cargo:warning=Not in a git repository; unable to embed git revision");
+      return Ok(String::new());
+    }
+    Err(err) => {
+      println!(
+        "cargo:warning=Failed to invoke `git`; unable to embed git revision: {}",
+        err
+      );
+      return Ok(String::new());
+    }
+  }
+
+  let local_changes = git_output(&["status", "--porcelain", "--untracked-files=no"])?;
+  let modified = !local_changes.is_empty();
+  let revision = git_output(&["rev-parse", "--short", "HEAD"])?;
+  let suffix = format!(" ({}{})", revision.trim(), if modified { "+" } else { "" });
```

Comment:
I’d like to have an indicator for builds from a tag.  Otherwise I always have to look up the revision for a tag to determine whether a release was built.  I could think of these options:
- If `git describe --exact-match` equals `v${CARGO_PKG_VERSION}`, use `"tag"` instead of `revision`:
  ```
  nitrocli 0.4.1 (tag+) using libnitrokey v3.6
  ```
- Use `git describe --always` instead of `git rev-parse --short HEAD`:
    ```
  nitrocli 0.4.1 (v0.4.1) using libnitrokey v3.6
  nitrocli 0.4.1 (v0.4.1-12-gad55ab2) using libnitrokey v3.6
  ```

Reply:
I am embedding the tag name now, if there is one. I think that's the most readable and useful.